pkg/dcgm/field_values.go

```diff
@@ -0,0 +1,124 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dcgm
+
+/*
+#include "dcgm_agent.h"
+#include "dcgm_structs.h"
+#include "field_values_cb.h"
+extern int go_dcgmFieldValueEntityEnumeration(dcgm_field_entity_group_t entityGroupId,
+            dcgm_field_eid_t entityId,
+            dcgmFieldValue_v1 *values,
+            int numValues,
+            void *userData);
+*/
+import "C"
+import (
+	"fmt"
+	"sync"
+	"time"
+	"unsafe"
+)
+
+// FieldValueEntity represents a field value entity
+type FieldValueEntity struct {
+	ts    int64
+	value [4096]byte
+}
+
+// AsInt64 returns field value as int64
+func (v FieldValueEntity) AsInt64() int64 {
+	ptr := (*C.int64_t)(unsafe.Pointer(&v.value[0]))
+	return int64(*ptr)
+}
+
+// FieldValueRecords represents field value records output in the format: [entityGroupId][entityId][fieldId][]FieldValue
+type FieldValues map[FieldEntityGroup]map[uint]map[int][]FieldValueEntity
+
+type callback struct {
+	mu     sync.Mutex
+	Values FieldValues
+}
+
+func (cb *callback) processValues(entityGroupID FieldEntityGroup, entityID uint, values []C.dcgmFieldValue_v1) {
+	if cb.Values == nil {
+		cb.Values = FieldValues{}
+	}
+
+	if _, exists := cb.Values[entityGroupID]; !exists {
+		cb.Values[entityGroupID] = map[uint]map[int][]FieldValueEntity{}
+	}
+
+	if _, exists := cb.Values[entityGroupID][entityID]; !exists {
+		cb.Values[entityGroupID][entityID] = map[int][]FieldValueEntity{}
+	}
+
+	cb.mu.Lock()
+
+	for _, v := range values {
+		if v.status > -1 {
+			cb.Values[entityGroupID][entityID][int(v.fieldId)] = append(cb.Values[entityGroupID][entityID][int(v.fieldId)],
+				FieldValueEntity{
+					ts:    int64(v.ts),
+					value: v.value,
+				})
+		}
+	}
+
+	cb.mu.Unlock()
+}
+
+//export go_dcgmFieldValueEntityEnumeration
+func go_dcgmFieldValueEntityEnumeration(
+	entityGroupID C.dcgm_field_entity_group_t,
+	entityID C.dcgm_field_eid_t,
+	values *C.dcgmFieldValue_v1,
+	numValues C.int,
+	userData unsafe.Pointer) C.int {
+	valuesSlice := (*[1 << 30]C.dcgmFieldValue_v1)(unsafe.Pointer(values))[0:numValues]
+	processor := (*callback)(unsafe.Pointer(userData))
+	processor.processValues(FieldEntityGroup(entityGroupID), uint(entityID), valuesSlice)
+	return 0
+}
+
+// GetValuesSince reads all field values that have updated since a given timestamp
+//
+// GPUGroupID Identifier for a group of GPUs. Look at CreateGroup for details on creating the group. Alternatively, pass in the group id as
+// DCGM_GROUP_ALL_GPUS to perform operation on all the GPUs or
+// DCGM_GROUP_ALL_NVSWITCHES to perform the operation on all NvSwitches.
+// fieldGroup Group of Fields to return data
+// since is a time to request values since. zero value (0) = request all data
+func GetValuesSince(GPUGroupID uint, fieldGroup FieldHandle, since time.Time) (FieldValues, time.Time, error) {
```

Comment:
Please document the return values as well. In particular, what time.Time in the return values stands for.